File: sdks/typescript/src/resources/goals.ts
import type { AegisClient } from "../client";
import type {
  Goal,
  GoalCreate,
  GoalUpdate,
  Milestone,
  MilestoneCreate,
  MilestoneUpdate,
  DeleteResponse,
} from "../types";

export class GoalResource {
  private client: AegisClient;

  constructor(client: AegisClient) {
    this.client = client;
  }

  // ─── Goals ───────────────────────────────────────────────────────────────

  /** List all goals. */
  async list(): Promise<Goal[]> {
    return this.client.get<Goal[]>("/api/goals/");
  }

  /** Create a new goal. */
  async create(data: GoalCreate): Promise<Goal> {
    return this.client.post<Goal>("/api/goals/", data);
  }

  /** Get a single goal by ID. */
  async get(id: number): Promise<Goal> {
    return this.client.get<Goal>(`/api/goals/${id}`);
  }

  /** Update an existing goal. */
  async update(id: number, data: GoalUpdate): Promise<Goal> {
    return this.client.put<Goal>(`/api/goals/${id}`, data);
  }

  /** Delete a goal. */
  async delete(id: number): Promise<DeleteResponse> {
    return this.client.delete<DeleteResponse>(`/api/goals/${id}`);
  }

  // ─── Milestones ──────────────────────────────────────────────────────────

  /** Create a milestone for a goal. */
  async createMilestone(data: MilestoneCreate): Promise<Milestone> {
    return this.client.post<Milestone>("/api/milestones/", data);
  }

  /** Update an existing milestone. */
  async updateMilestone(id: number, data: MilestoneUpdate): Promise<Milestone> {
    return this.client.put<Milestone>(`/api/milestones/${id}`, data);
  }

  /** Delete a milestone. */
  async deleteMilestone(id: number): Promise<DeleteResponse> {
    return this.client.delete<DeleteResponse>(`/api/milestones/${id}`);
  }
}

File: sdks/go/goals.go
package aegis

import "fmt"

// GoalService handles operations on the /api/goals and /api/milestones endpoints.
type GoalService struct {
	client *Client
}

// List returns all goals.
func (s *GoalService) List() ([]Goal, error) {
	var goals []Goal
	err := s.client.get("/api/goals/", &goals)
	return goals, err
}

// Create adds a new goal.
func (s *GoalService) Create(goal GoalCreate) (*Goal, error) {
	var created Goal
	err := s.client.post("/api/goals/", &goal, &created)
	return &created, err
}

// Get retrieves a single goal by ID.
func (s *GoalService) Get(id int) (*Goal, error) {
	var goal Goal
	err := s.client.get(fmt.Sprintf("/api/goals/%d", id), &goal)
	return &goal, err
}

// Update modifies an existing goal by ID.
func (s *GoalService) Update(id int, goal GoalUpdate) (*Goal, error) {
	var updated Goal
	err := s.client.put(fmt.Sprintf("/api/goals/%d", id), &goal, &updated)
	return &updated, err
}

// Delete removes a goal by ID.
func (s *GoalService) Delete(id int) (*MessageResponse, error) {
	var resp MessageResponse
	err := s.client.delete(fmt.Sprintf("/api/goals/%d", id), &resp)
	return &resp, err
}

// ---------- Milestones ----------

// ListMilestones returns all milestones.
func (s *GoalService) ListMilestones() ([]Milestone, error) {
	var milestones []Milestone
	err := s.client.get("/api/milestones/", &milestones)
	return milestones, err
}

// CreateMilestone adds a new milestone.
func (s *GoalService) CreateMilestone(ms MilestoneCreate) (*Milestone, error) {
	var created Milestone
	err := s.client.post("/api/milestones/", &ms, &created)
	return &created, err
}

// UpdateMilestone modifies an existing milestone by ID.
func (s *GoalService) UpdateMilestone(id int, ms MilestoneUpdate) (*Milestone, error) {
	var updated Milestone
	err := s.client.put(fmt.Sprintf("/api/milestones/%d", id), &ms, &updated)
	return &updated, err
}

// DeleteMilestone removes a milestone by ID.
func (s *GoalService) DeleteMilestone(id int) (*MessageResponse, error) {
	var resp MessageResponse
	err := s.client.delete(fmt.Sprintf("/api/milestones/%d", id), &resp)
	return &resp, err
}
